### PR TITLE
Added code for PRPN

### PR DESCRIPTION
**Pull request by phu-pmh**

### CODE REVIEW

**Comment by sleepinyourhat** [comment]
Thanks for setting this up! 

You've got some merge conflicts now that Anhad's PR is done. Update from master and add a note when this is ready to review.

File: README.md [hunk 1]
```diff
@@ -149,6 +149,7 @@ The base model class is a MultiTaskModel. To add another model, first add the cl
 Task-specific components include logistic regression and multi-layer perceptron for classification and regression tasks, and an RNN decoder with attention for sequence transduction tasks.
 To see the full set of available params, see [config/defaults.conf](config/defaults.conf). For a list of options affecting the execution pipeline (which configuration file to use, whether to enable remote logging or tensorboard, etc.), see the arguments section in [main.py](main.py).
 
+<<<<<<< HEAD
```
Comment (W4ngatang) [inline README.md]: fix

Comment (phu-pmh) [reply README.md]: Yes, I can do that. I'll talk with Anhad first about ONLSTM and will ask you.

**Comment by phu-pmh** [comment]
Fixed merge conflicts with the master

**Comment by sleepinyourhat** [comment]
Ready for review, or planning to split this up further?

**Comment by phu-pmh** [comment]
It's ready for review. No plan to split further for now.

File: .gitignore [hunk 2]
```diff
@@ -6,3 +6,13 @@ glue_data/
...
+*.tgz
+*.sbatch
+commoncrawl.*
+subword-nmt/
+mosesdecoder/
+test-full/*
+prpn_logs
+prpn_logs/*
+sl*.out
+sl*.err
```
Comment (sleepinyourhat) [inline .gitignore]: Changing this file isn't a good idea unless there's a really clear need—leaving unexpected things in gitignore is a recipe for broken PRs later, and you shouldn't need to: Just use a working directory that's outside jiant/.

File: config/defaults.conf [hunk 3]
```diff
@@ -243,6 +243,7 @@ onlstm_tying = 0  // Language Modeling tying of weights.
 
 force_include_wsj_vocabulary = 0  // Set if using for PTB parsing task. Makes sure to include WSJ vocabulary.
 
+
 // Sentence Encoder //
 sent_enc = rnn  // Type of sentence encoder: 'bow', 'rnn' (for LSTM), or 'transformer'
```
Comment (sleepinyourhat) [inline config/defaults.conf]: Mention PRPN as an option somewhere in defaults.

File: README.md [hunk 4]
```diff
@@ -280,4 +280,4 @@ Post an issue here on GitHub if you have any problems, and create a pull request
 
 It seems like my preproc/{task}__{split}.data has nothing in it!
 
-This probably means that you probably ran the script before downloading the data for that task. Thus, delete the file from preproc and then run main.py again to build the data splits from scratch.
+This probably means that you probably ran the script before downloading the data for that task. Thus, delete the file from preproc and then run main.py again to build the data splits from scratch.
```
Comment (sleepinyourhat) [inline README.md]: Try to remove files from the PR when they're not really changing. Includes this file and others below.

File: config/prpn.conf [hunk 5]
```diff
@@ -0,0 +1,40 @@
...
+NFS_PROJECT_PREFIX=/beegfs/pmh330/jiant-prpn-outs
+project_dir = ${NFS_PROJECT_PREFIX}  // The base directory for model output. This default is meant for runs
+                                     // on the JSALT Google Cloud Platform (GCP) setup, and you'll need to
+                                     // override it elsewhere.
+exp_name = prpn-wsj  // Experiment name, will be a subdirectory of project_dir.
+                                   // This directory will contain all run directories, a results summary file,
+                                   // and preprocessed data files.
+run_name = prpn-0  // Run name, will be a subdirectory of exp_name. This directory will contain logs,
+                       // checkpoints, and model predictions.
+JIANT_DATA_DIR=/scratch/pmh330/jiant-data/
```
Comment (sleepinyourhat) [inline config/prpn.conf]: Don't override this—use path_config.sh (or similar) instead.

Comment (sleepinyourhat) [reply config/prpn.conf]: Same for NFS and project_dir above.

Comment (sleepinyourhat) [reply config/prpn.conf]: And data_dir below.

File: config/prpn.conf [hunk 6]
```diff
@@ -0,0 +1,40 @@
...
+data_dir = ${JIANT_DATA_DIR}  // Base directory in which to look for raw data subdirectories. This will often be
+FASTTEXT_MODEL_FILE=""
+word_embs_file = ""
```
Comment (sleepinyourhat) [inline config/prpn.conf]: You shouldn't need to set these here.

File: config/prpn.conf [hunk 7]
```diff
@@ -0,0 +1,40 @@
...
+bidirectional = 0
+pretrain_tasks = "wsj"
+target_tasks = "wsj"
+do_pretrain = 1
+do_target_task_training = 1
+do_full_eval = 1
+allow_reuse_of_pretraining_parameters = 1
+word_embs = "scratch"
+sent_enc = "prpn"
+d_hid = 400  // default ON-LSTM paper hyperparameter
```
Comment (sleepinyourhat) [inline config/prpn.conf]: Why are the comments talking about ON-LSTM here? Intentional?

File: config/defaults.conf (hunk 3 above)
Comment (sleepinyourhat) [reply config/defaults.conf]: (and ON-LSTM, if it's missing...)

File: config/prpn.conf [hunk 8]
```diff
@@ -0,0 +1,40 @@
...
+n_layers_enc = 2  // default ON-LSTM paper hyperparameter
+dropout = 0.2  // default ON-LSTM paper hyperparameter
+d_word = 200  // default ON-LSTM paper hyperparameter
+n_slots=15
```
Comment (sleepinyourhat) [inline config/prpn.conf]: Say what this is...

File: environment.yml [hunk 9]
```diff
@@ -2,6 +2,7 @@ name: jiant
 channels:
     - pytorch
 dependencies:
+    - pip
```
Comment (sleepinyourhat) [inline environment.yml]: Intentional change?

File: src/models.py [hunk 10]
```diff
@@ -29,8 +29,8 @@
 
 from .preprocess import parse_task_list_arg, get_tasks
 
-from .tasks.tasks import CCGTaggingTask, ClassificationTask, CoLATask, CoLAAnalysisTask, \
-    GroundedSWTask, GroundedTask, MultiNLIDiagnosticTask, PairClassificationTask, \
+from .tasks.tasks import CCGTaggingTask, ClassificationTask, CoLATask, GroundedSWTask, \
+    GroundedTask, MultiNLIDiagnosticTask, PairClassificationTask, \
```
Comment (sleepinyourhat) [inline src/models.py]: This is an enormous change, and most of it doesn't have any obvious connection to PRPN. Trim this down to something more focused and we can review that.

Comment (sleepinyourhat) [reply src/models.py]: For the most part, you should only need to change sections of code that were changed in the ON-LSTM PR.

File: src/modules/modules.py [hunk 11]
```diff
@@ -270,6 +276,53 @@ def forward(self, sent, task):
...
+        initializer(self)
+
+    def get_input_dim(self):
+        return self.prpnlayer.ninp
+
+    def get_output_dim(self):
+        return self.prpnlayer.ninp
+
+
+class ONLSTMSentEncoder(Model):
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: Why are you changing this?

File: src/modules/prpn/LSTMCell.py [hunk 12]
```diff
@@ -0,0 +1,47 @@
+import torch.nn.functional as F
+import torch.nn as nn
+from torch.nn.modules.rnn import *
+
```
Comment (sleepinyourhat) [inline src/modules/prpn/LSTMCell.py]: Add a comment explaining what this file is for. If any of it is reused from PRPN, say so/give credit. If any of it already exists in AllenNLP or PyTorch, either switch to that version, or at least make it clear that you need to use this version to match the PRPN paper in some way.

Comment (sleepinyourhat) [reply src/modules/prpn/LSTMCell.py]: Same with the rest of this directory.

File: src/modules/prpn/blocks.py [hunk 13]
```diff
@@ -0,0 +1,48 @@
...
+
+def stick_breaking(logits):
+    e = F.sigmoid(logits)
+    z = (1 - e).cumprod(dim=1)
+    p = torch.cat([e.narrow(1, 0, 1), e[:, 1:] * z[:, :-1]], dim=1)
+
+    return p
+
+
+def softmax(x, mask=None):
```
Comment (sleepinyourhat) [inline src/modules/prpn/blocks.py]: This is a bit odd. There is _definitely_ an efficient implementation of this built into PyTorch...

File: src/preprocess.py [hunk 14]
```diff
@@ -150,7 +147,7 @@ def _counter_iter(elems):
 
 
 def _find_cached_file(exp_dir: str, global_exp_cache_dir: str,
-                      relative_path: str, log_prefix: str = "") -> bool:
+                      relative_path: str, log_prefix: str="") -> bool:
```
Comment (sleepinyourhat) [inline src/preprocess.py]: autopep8 for style.

Comment (sleepinyourhat) [reply src/preprocess.py]: (Search the repo for details.)

File: src/preprocess.py [hunk 15]
```diff
@@ -39,11 +39,8 @@
 from .tasks import ALL_GLUE_TASKS, ALL_NLI_PROBING_TASKS, ALL_TARG_VOC_TASKS
 from .tasks.mt import MTTask
 
-
```
Comment (sleepinyourhat) [inline src/preprocess.py]: Why do you need to make changes here?

File: src/tasks/lm.py [hunk 16]
```diff
@@ -79,7 +76,7 @@ def load_data(self, path):
                 toks = row.strip()
                 if not toks:
                     continue
-                yield process_sentence(self._tokenizer_name, toks, self.max_seq_len)
```
Comment (sleepinyourhat) [inline src/tasks/lm.py]: Why do we need this change?

File: src/tasks/lm_parsing.py [hunk 17]
```diff
@@ -13,6 +13,7 @@
 # Fields for instance processing
```
Comment (sleepinyourhat) [inline src/tasks/lm_parsing.py]: Revert these changes.

**Comment by sleepinyourhat** [comment]
This wasn't ready. :/

In general:
- Make sure all changes in a PR directly relate to the thing you're adding. If you're adding more than one feature that can be added separately, add them separately.
- Review the PR yourself (in the 'Files Changed') view before you ask others for review, to make sure that everything makes sense.

File: src/modules/prpn/blocks.py (hunk 13 above)
Comment (phu-pmh) [reply src/modules/prpn/blocks.py]: They modify the built-in softmax to include masking for self-attention.

File: README.md [hunk 18]
```diff
@@ -150,6 +150,8 @@ Task-specific components include logistic regression and multi-layer perceptron
 To see the full set of available params, see [config/defaults.conf](config/defaults.conf). For a list of options affecting the execution pipeline (which configuration file to use, whether to enable remote logging or tensorboard, etc.), see the arguments section in [main.py](main.py).
 
 To use the ON-LSTM sentence encoder from [Ordered Neurons: Integrating Tree Structures into Recurrent Neural Networks](https://arxiv.org/abs/1810.09536), set ``sent_enc = onlstm``. To re-run experiments from the paper on WSJ Language Modeling, use the configuration file [config/onlstm.conf](config/onlstm.conf). Specific ON-LSTM modules use code from the [Github](https://github.com/yikangshen/Ordered-Neurons) implementation of the paper.
+
+To use the PRPN sentence encoder from [***Neural language modeling by jointly learning syntax and lexicon***](https://arxiv.org/abs/1711.02013), set ``sent_enc=prpn``. To re-run experiments from the paper on WSJ Language Modeling, use the configuration file [config/prpn.conf](config/prpn.conf).  
 ### Transformers
```
Comment (sleepinyourhat) [inline README.md]: Line break here.

Comment (sleepinyourhat) [reply README.md]: Give credit here for the code. (See ON-LSTM.)

File: src/models.py [hunk 19]
```diff
@@ -702,6 +715,8 @@ def forward(self, task, batch, predict=False):
         elif isinstance(task, LanguageModelingTask):
             if isinstance(self.sent_encoder._phrase_layer, ONLSTMStack):
                 out = self._lm_only_lr_forward(batch, task)
+            elif isinstance(self.sent_encoder._phrase_layer, PRPN):
```
Comment (sleepinyourhat) [inline src/models.py]: You can save a line by combining this with the previous 'if'.

File: src/modules/modules.py [hunk 20]
```diff
@@ -191,6 +192,9 @@ def forward(self, sent, task, reset=True):
                 # The ONLSTMStack takes the raw words as input and computes embeddings separately.
                 sent_enc, _ = self._phrase_layer(torch.transpose(sent["words"], 0, 1), sent_lstm_mask)
                 sent_enc = torch.transpose(sent_enc, 0, 1)
+            elif isinstance(self._phrase_layer, PRPN):
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: Don't duplicate code.

File: src/modules/modules.py [hunk 21]
```diff
@@ -213,8 +217,10 @@ def forward(self, sent, task, reset=True):
             else:
                 sent_enc = torch.cat([sent_enc, skip_vec], dim=-1)
 
+
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: Revert this—it's not a good idea to change code that isn't crucial for the PR, and two blank lines within a method is pretty unusual.

File: src/modules/modules.py [hunk 22]
```diff
@@ -270,6 +276,36 @@ def forward(self, sent, task):
         word_mask = util.get_text_field_mask(sent).float()
         return word_embs, word_mask  # need to get # nonzero elts
 
+class PRPNPhraseLayer(Model):
+    ''' PRPN sentence encoder '''
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: Try not to use 'sentence encoder' here, since this object isn't of type SentenceEncoder.

Comment (sleepinyourhat) [reply src/modules/modules.py]: (I know ON-LSTM does this below. Consider fixing that too, while you're at it.)

File: src/modules/prpn/LSTMCell.py [hunk 23]
```diff
@@ -0,0 +1,52 @@
+'''
+ LSTMCell used in the Reading Network of PRPN
```
Comment (sleepinyourhat) [inline src/modules/prpn/LSTMCell.py]: Fix space.

File: src/modules/prpn/LSTMCell.py [hunk 24]
```diff
@@ -0,0 +1,52 @@
...
+Reference: Parsing-Reading-Predict Networks (PRPN; Shen et al., 2018)
+Modules re-used from: https://github.com/yikangshen/PRPN
```
Comment (sleepinyourhat) [inline src/modules/prpn/LSTMCell.py]: This is a bit vague. All modules? All code? Parts of some specific modules?

File: src/modules/prpn/PRPN.py [hunk 25]
```diff
@@ -0,0 +1,142 @@
+'''
+Code for Parsing-Reading-Predict Networks (PRPN; Shen et al., 2018)
+Modules re-used from: https://github.com/yikangshen/PRPN
```
Comment (sleepinyourhat) [inline src/modules/prpn/PRPN.py]: Same comment as above. It should be clear which lines are exactly their code, and which lines are your edits of their code.

Comment (sleepinyourhat) [reply src/modules/prpn/PRPN.py]: And same in other similar files.

**Comment by phu-pmh** [comment]
Updated the code based on the comments.

File: src/modules/modules.py [hunk 26]
```diff
@@ -270,9 +273,39 @@ def forward(self, sent, task):
         word_mask = util.get_text_field_mask(sent).float()
         return word_embs, word_mask  # need to get # nonzero elts
 
+class PRPNPhraseLayer(Model):
+    ''' PRPN phrase layer '''
```
Comment (sleepinyourhat) [inline src/modules/modules.py]: I was hoping for something that would actually say what PRPN (and ON) are, since most people who use this file will have never seen LTL.

File: src/modules/prpn/PRPN.py (hunk 25 above)
Comment (sleepinyourhat) [reply src/modules/prpn/PRPN.py]: This still seems a bit vague, which is really, really dangerous. How about something like: 'This file is a version of a class from [URL], modified to integrate with jiant.'

Comment (sleepinyourhat) [reply src/modules/prpn/PRPN.py]: The other files look good, though!

**Comment by phu-pmh** [comment]
Updated the code based on the comments. I've also tested it and it runs fine.
